com.hazelcast.jet.impl.util.AsyncMapWriterTest.when_writeErrorOnlyOnce_then_retrySuccess

```
Error Message
expected:<0> but was:<null>
Stacktrace
java.lang.AssertionError: expected:<0> but was:<null>
	at com.hazelcast.jet.impl.util.AsyncMapWriterTest.when_writeErrorOnlyOnce_then_retrySuccess(AsyncMapWriterTest.java:260)
```

https://hazelcast-l337.ci.cloudbees.com/view/Jet/job/jet/com.hazelcast.jet$hazelcast-jet-core/825/testReport/junit/com.hazelcast.jet.impl.util/AsyncMapWriterTest/when_writeErrorOnlyOnce_then_retrySuccess/

--- a/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/util/AsyncMapWriter.java
+++ b/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/util/AsyncMapWriter.java
@@ -119,7 +119,7 @@ public class AsyncMapWriter {
      */
     public boolean tryFlushAsync(CompletableFuture<Void> completionFuture) {
         Map<Address, List<Integer>> memberPartitionsMap = partitionService.getMemberPartitionsMap();
-        AtomicInteger doneLatch = new AtomicInteger(0);
+        AtomicInteger pendingOps = new AtomicInteger(0);
         List<PartitionOpBuilder> ops = memberPartitionsMap.entrySet()
                                                           .stream()
                                                           .map(e -> opForMember(e.getKey(), e.getValue(), outputBuffers))
@@ -131,14 +131,14 @@ public class AsyncMapWriter {
             return true;
         }
 
-        if (!invokeOnCluster(ops, doneLatch, completionFuture, true)) {
+        if (!invokeOnCluster(ops, pendingOps, completionFuture, true)) {
             return false;
         }
         resetBuffers();
         return true;
     }
 
-    private boolean tryRetry(int[] partitions, MapEntries[] entriesPerPtion, AtomicInteger doneLatch,
+    private boolean tryRetry(int[] partitions, MapEntries[] entriesPerPtion, AtomicInteger pendingOps,
                              CompletableFuture<Void> completionFuture) {
         assert partitions.length == entriesPerPtion.length;
         Map<Address, Entry<List<Integer>, List<MapEntries>>> addrToEntries = new HashMap<>();
@@ -164,7 +164,7 @@ public class AsyncMapWriter {
                     return h;
                 }).collect(Collectors.toList());
 
-        return invokeOnCluster(retryOps, doneLatch, completionFuture, false);
+        return invokeOnCluster(retryOps, pendingOps, completionFuture, false);
     }
 
     private PartitionOpBuilder opForMember(Address member, List<Integer> partitions, MapEntries[] partitionToEntries) {
@@ -197,14 +197,14 @@ public class AsyncMapWriter {
     }
 
     private boolean invokeOnCluster(List<PartitionOpBuilder> opBuilders,
-                                    AtomicInteger doneLatch,
+                                    AtomicInteger pendingOps,
                                     CompletableFuture<Void> completionFuture,
                                     boolean shouldRetry) {
         Preconditions.checkFalse(opBuilders.isEmpty(), "opBuilders is empty");
         if (!tryIncrement(numConcurrentOps, opBuilders.size(), MAX_PARALLEL_ASYNC_OPS)) {
             return false;
         }
-        doneLatch.addAndGet(opBuilders.size());
+        pendingOps.addAndGet(opBuilders.size());
         for (PartitionOpBuilder builder : opBuilders) {
             ExecutionCallback<PartitionResponse> callback = callbackOf(r -> {
                 numConcurrentOps.decrementAndGet();
@@ -239,12 +239,12 @@ public class AsyncMapWriter {
                     final Throwable originalErr = error;
                     executionService.schedule(() -> {
                         try {
-                            if (!tryRetry(partitions, entries, doneLatch, completionFuture)) {
+                            // We should not handle pendingOps getting to 0 here, it will be increased again by
+                            // the retry operations.
+                            pendingOps.decrementAndGet();
+                            if (!tryRetry(partitions, entries, pendingOps, completionFuture)) {
                                 completionFuture.completeExceptionally(originalErr);
-                                return;
                             }
-                            int value = doneLatch.decrementAndGet();
-                            assert value > 0 : "value=" + value;
                         } catch (Exception e) {
                             logger.severe("Exception during retry", e);
                             completionFuture.completeExceptionally(originalErr);
@@ -252,7 +252,7 @@ public class AsyncMapWriter {
                     }, TRY_PAUSE_MILLIS, TimeUnit.MILLISECONDS);
                     return;
                 }
-                if (doneLatch.decrementAndGet() == 0) {
+                if (pendingOps.decrementAndGet() == 0) {
                     completionFuture.complete(null);
                 }
 
@@ -261,7 +261,7 @@ public class AsyncMapWriter {
                 if (throwable instanceof RetryableException) {
                     // the whole operation to the member failed, so we need to retry
                     // all of the partitions in the operation
-                    if (!tryRetry(builder.partitions, builder.entries, doneLatch, completionFuture)) {
+                    if (!tryRetry(builder.partitions, builder.entries, pendingOps, completionFuture)) {
                         completionFuture.completeExceptionally(throwable);
                     }
                 } else {

--- a/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/util/AsyncMapWriter.java
+++ b/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/util/AsyncMapWriter.java
@@ -242,6 +242,10 @@ public class AsyncMapWriter {
                             // We should not handle pendingOps getting to 0 here, it will be increased again by
                             // the retry operations.
                             pendingOps.decrementAndGet();
+                            // TODO do more robust retry
+                            // On second try we should do individual partition-specific ops that can be retried on
+                            // different members as the partition migrates, not a PartitionIteratingOp.
+                            // See InvokeOnPartitions.retryFailedPartitions.
                             if (!tryRetry(partitions, entries, pendingOps, completionFuture)) {
                                 completionFuture.completeExceptionally(originalErr);
                             }

--- a/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/util/AsyncMapWriter.java
+++ b/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/util/AsyncMapWriter.java
@@ -34,6 +34,7 @@ import com.hazelcast.spi.impl.operationservice.impl.operations.PartitionIteratin
 import com.hazelcast.spi.partition.IPartitionService;
 import com.hazelcast.spi.serialization.SerializationService;
 import com.hazelcast.util.CollectionUtil;
+import com.hazelcast.util.Preconditions;
 
 import java.util.ArrayList;
 import java.util.Arrays;
@@ -118,20 +119,27 @@ public class AsyncMapWriter {
      */
     public boolean tryFlushAsync(CompletableFuture<Void> completionFuture) {
         Map<Address, List<Integer>> memberPartitionsMap = partitionService.getMemberPartitionsMap();
+        AtomicInteger doneLatch = new AtomicInteger(0);
         List<PartitionOpBuilder> ops = memberPartitionsMap.entrySet()
                                                           .stream()
                                                           .map(e -> opForMember(e.getKey(), e.getValue(), outputBuffers))
                                                           .filter(Objects::nonNull)
                                                           .collect(Collectors.toList());
 
-        if (!invokeOnCluster(ops, completionFuture, true)) {
+        if (ops.isEmpty()) {
+            completionFuture.complete(null);
+            return true;
+        }
+
+        if (!invokeOnCluster(ops, doneLatch, completionFuture, true)) {
             return false;
         }
         resetBuffers();
         return true;
     }
 
-    private boolean tryRetry(int[] partitions, MapEntries[] entriesPerPtion, CompletableFuture<Void> completionFuture) {
+    private boolean tryRetry(int[] partitions, MapEntries[] entriesPerPtion, AtomicInteger doneLatch,
+                             CompletableFuture<Void> completionFuture) {
         assert partitions.length == entriesPerPtion.length;
         Map<Address, Entry<List<Integer>, List<MapEntries>>> addrToEntries = new HashMap<>();
         for (int index = 0; index < partitions.length; index++) {
@@ -151,12 +159,12 @@ public class AsyncMapWriter {
                 .map(e -> {
                     PartitionOpBuilder h = new PartitionOpBuilder(e.getKey());
                     List<MapEntries> entries = e.getValue().getValue();
-                    h.entries = entries.toArray(new MapEntries[entries.size()]);
+                    h.entries = entries.toArray(new MapEntries[0]);
                     h.partitions = CollectionUtil.toIntArray(e.getValue().getKey());
                     return h;
                 }).collect(Collectors.toList());
 
-        return invokeOnCluster(retryOps, completionFuture, false);
+        return invokeOnCluster(retryOps, doneLatch, completionFuture, false);
     }
 
     private PartitionOpBuilder opForMember(Address member, List<Integer> partitions, MapEntries[] partitionToEntries) {
@@ -184,23 +192,19 @@ public class AsyncMapWriter {
         return builder;
     }
 
-
     private void resetBuffers() {
         Arrays.fill(outputBuffers, null);
     }
 
     private boolean invokeOnCluster(List<PartitionOpBuilder> opBuilders,
+                                    AtomicInteger doneLatch,
                                     CompletableFuture<Void> completionFuture,
                                     boolean shouldRetry) {
-        if (opBuilders.isEmpty()) {
-            completionFuture.complete(null);
-            return true;
-        }
-
+        Preconditions.checkFalse(opBuilders.isEmpty(), "opBuilders is empty");
         if (!tryIncrement(numConcurrentOps, opBuilders.size(), MAX_PARALLEL_ASYNC_OPS)) {
             return false;
         }
-        AtomicInteger doneLatch = new AtomicInteger(opBuilders.size());
+        doneLatch.addAndGet(opBuilders.size());
         for (PartitionOpBuilder builder : opBuilders) {
             ExecutionCallback<PartitionResponse> callback = callbackOf(r -> {
                 numConcurrentOps.decrementAndGet();
@@ -230,15 +234,17 @@ public class AsyncMapWriter {
                     }
 
                     // retry once
-                    final MapEntries[] entries = failedEntries.toArray(new MapEntries[failedEntries.size()]);
+                    final MapEntries[] entries = failedEntries.toArray(new MapEntries[0]);
                     final int[] partitions = toIntArray(failedPartitions);
                     final Throwable originalErr = error;
                     executionService.schedule(() -> {
                         try {
-                            if (!tryRetry(partitions, entries, completionFuture)) {
+                            if (!tryRetry(partitions, entries, doneLatch, completionFuture)) {
                                 completionFuture.completeExceptionally(originalErr);
+                                return;
                             }
-
+                            int value = doneLatch.decrementAndGet();
+                            assert value > 0 : "value=" + value;
                         } catch (Exception e) {
                             logger.severe("Exception during retry", e);
                             completionFuture.completeExceptionally(originalErr);
@@ -255,7 +261,7 @@ public class AsyncMapWriter {
                 if (throwable instanceof RetryableException) {
                     // the whole operation to the member failed, so we need to retry
                     // all of the partitions in the operation
-                    if (!tryRetry(builder.partitions, builder.entries, completionFuture)) {
+                    if (!tryRetry(builder.partitions, builder.entries, doneLatch, completionFuture)) {
                         completionFuture.completeExceptionally(throwable);
                     }
                 } else {
@@ -299,4 +305,3 @@ public class AsyncMapWriter {
         }
     }
 }
-

--- a/hazelcast-jet-core/src/test/java/com/hazelcast/jet/core/CancellationTest.java
+++ b/hazelcast-jet-core/src/test/java/com/hazelcast/jet/core/CancellationTest.java
@@ -259,13 +259,26 @@ public class CancellationTest extends JetTestSupport {
     }
 
     @Test
-    public void when_shutdown_then_jobFuturesCanceled() {
+    public void when_shutdownGracefully_then_jobFuturesCanceled() {
+        when_shutdown_then_jobFuturesCanceled(true);
+    }
+
+    @Test
+    public void when_shutdownForcefully_then_jobFuturesCanceled() {
+        when_shutdown_then_jobFuturesCanceled(false);
+    }
+
+    private void when_shutdown_then_jobFuturesCanceled(boolean graceful) {
         JetInstance jet = newInstance();
         DAG dag = new DAG();
         dag.newVertex("blocking", BlockingProcessor::new).localParallelism(1);
         jet.newJob(dag);
         assertTrueEventually(() -> assertTrue(BlockingProcessor.hasStarted), 3);
-        jet.shutdown();
+        if (graceful) {
+            jet.shutdown();
+        } else {
+            jet.getHazelcastInstance().shutdown();
+        }
         assertBlockingProcessorEventuallyNotRunning();
     }
 

--- a/hazelcast-jet-core/src/test/java/com/hazelcast/jet/core/ManualRestartTest.java
+++ b/hazelcast-jet-core/src/test/java/com/hazelcast/jet/core/ManualRestartTest.java
@@ -136,7 +136,7 @@ public class ManualRestartTest extends JetTestSupport {
             job.join();
             fail();
         } catch (CancellationException ignored) {
-            System.out.println("Cancellation exception caught");
+            logger.info("Cancellation exception caught");
         }
 
         // Then, the job cannot restart

--- a/hazelcast-jet-core/src/test/java/com/hazelcast/jet/impl/connector/StreamEventJournalPTest.java
+++ b/hazelcast-jet-core/src/test/java/com/hazelcast/jet/impl/connector/StreamEventJournalPTest.java
@@ -176,7 +176,7 @@ public class StreamEventJournalPTest extends JetTestSupport {
         List<Entry> snapshotItems = new ArrayList<>();
         outbox.drainSnapshotQueueAndReset(snapshotItems, false);
 
-        System.out.println("Restoring journal");
+        logger.info("Restoring journal");
         // restore from snapshot
         assertRestore(snapshotItems);
     }

--- a/hazelcast-jet-core/src/test/java/com/hazelcast/jet/impl/connector/StreamFilesP_integrationTest.java
+++ b/hazelcast-jet-core/src/test/java/com/hazelcast/jet/impl/connector/StreamFilesP_integrationTest.java
@@ -286,9 +286,9 @@ public class StreamFilesP_integrationTest extends JetTestSupport {
 
     private void finishDirectory(Future<Void> jobFuture, File ... files) throws Exception {
         for (File file : files) {
-            System.out.println("deleting " + file + "...");
+            logger.info("deleting " + file + "...");
             assertTrueEventually(() -> assertTrue("Failed to delete " + file, file.delete()));
-            System.out.println("deleted " + file);
+            logger.info("deleted " + file);
         }
         assertTrueEventually(() -> assertTrue("Failed to delete " + directory, directory.delete()));
         assertTrueEventually(() -> assertTrue("job should complete eventually", jobFuture.isDone()));

--- a/hazelcast-jet-core/src/test/java/com/hazelcast/jet/impl/connector/StreamSocketPTest.java
+++ b/hazelcast-jet-core/src/test/java/com/hazelcast/jet/impl/connector/StreamSocketPTest.java
@@ -90,7 +90,7 @@ public class StreamSocketPTest extends JetTestSupport {
         // split at any place: between \r\n, between the bytes of utf-16 sequence etc
         byte[] inputBytes = input.getBytes(UTF_16);
         for (int splitIndex = 0; splitIndex < inputBytes.length; splitIndex++) {
-            System.out.println("--------- runTest(" + splitIndex + ") ---------");
+            logger.info("--------- runTest(" + splitIndex + ") ---------");
             runTest(inputBytes, splitIndex);
         }
     }

--- a/hazelcast-jet-core/src/test/java/com/hazelcast/jet/impl/connector/WriteBufferedPTest.java
+++ b/hazelcast-jet-core/src/test/java/com/hazelcast/jet/impl/connector/WriteBufferedPTest.java
@@ -91,23 +91,19 @@ public class WriteBufferedPTest extends JetTestSupport {
     @Test
     public void when_writeBufferedJobFailed_then_bufferDisposed() throws Exception {
         JetInstance instance = createJetMember();
-        try {
-            DAG dag = new DAG();
-            Vertex source = dag.newVertex("source", StuckForeverSourceP::new);
-            Vertex sink = dag.newVertex("sink", getLoggingBufferedWriter()).localParallelism(1);
+        DAG dag = new DAG();
+        Vertex source = dag.newVertex("source", StuckForeverSourceP::new);
+        Vertex sink = dag.newVertex("sink", getLoggingBufferedWriter()).localParallelism(1);
 
-            dag.edge(Edge.between(source, sink));
+        dag.edge(Edge.between(source, sink));
 
-            Job job = instance.newJob(dag);
-            // wait for the job to initialize
-            Thread.sleep(5000);
-            job.cancel();
+        Job job = instance.newJob(dag);
+        // wait for the job to initialize
+        Thread.sleep(5000);
+        job.cancel();
 
-            assertTrueEventually(() -> assertTrue("No \"dispose\", only: " + events, events.contains("dispose")), 60);
-            System.out.println(events);
-        } finally {
-            instance.shutdown();
-        }
+        assertTrueEventually(() -> assertTrue("No \"dispose\", only: " + events, events.contains("dispose")), 60);
+        System.out.println(events);
     }
 
     // returns a processor that will not write anywhere, just log the events

--- a/hazelcast-jet-core/src/test/java/com/hazelcast/jet/impl/util/AsyncMapWriterTest.java
+++ b/hazelcast-jet-core/src/test/java/com/hazelcast/jet/impl/util/AsyncMapWriterTest.java
@@ -227,7 +227,7 @@ public class AsyncMapWriterTest extends JetTestSupport {
         boolean flushed = writer.tryFlushAsync(future);
 
         // Then
-        assertTrue("tryFlush() returned false", flushed);
+        assertTrue("tryFlushAsync() returned false", flushed);
 
         Throwable actual = future.handle((r, e) -> e).join();
         assertNotNull("No exception was thrown", actual);
@@ -254,14 +254,14 @@ public class AsyncMapWriterTest extends JetTestSupport {
         boolean flushed = writer.tryFlushAsync(future);
 
         // Then
-        assertTrue("tryFlush() returned false", flushed);
+        assertTrue("tryFlushAsync() returned false", flushed);
         future.join();
         for (int i = 0; i < 100; i++) {
             assertEquals(i, map.get(i));
         }
     }
 
-    public static class AlwaysFailingMapStore extends AMapStore implements Serializable {
+    static class AlwaysFailingMapStore extends AMapStore implements Serializable {
 
         @Override
         public void store(Object o, Object o2) {
@@ -269,7 +269,7 @@ public class AsyncMapWriterTest extends JetTestSupport {
         }
     }
 
-    public static class RetryableMapStore extends AMapStore implements Serializable {
+    private static class RetryableMapStore extends AMapStore implements Serializable {
 
         private static volatile boolean failOnNext;
 

--- a/hazelcast-jet-core/src/test/java/com/hazelcast/jet/impl/util/AsyncMapWriterTest.java
+++ b/hazelcast-jet-core/src/test/java/com/hazelcast/jet/impl/util/AsyncMapWriterTest.java
@@ -56,7 +56,6 @@ public class AsyncMapWriterTest extends JetTestSupport {
     private IMapJet<Object, Object> map;
     private NodeEngineImpl nodeEngine;
 
-
     @Before
     public void setup() {
         JetConfig jetConfig = new JetConfig();

